lib/dice/src/lib.rs

```diff
@@ -0,0 +1,225 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#![cfg_attr(not(test), no_std)]
+
+use core::{mem, ptr, str::FromStr};
+use hkdf::Hkdf;
+use hubpack::SerializedSize;
+use salty::constants::SECRETKEY_SEED_LENGTH;
+use serde::{Deserialize, Serialize};
+use sha3::Sha3_256;
+use unwrap_lite::UnwrapLite;
+use zeroize::Zeroize;
+
+mod cert;
+pub use crate::cert::{AliasCert, CertError, DeviceIdCert};
+mod alias_cert_tmpl;
+mod deviceid_cert_tmpl;
+mod handoff;
+pub use crate::handoff::{AliasHandoff, Handoff};
+
+pub const SEED_LENGTH: usize = SECRETKEY_SEED_LENGTH;
+pub const SN_LENGTH: usize = deviceid_cert_tmpl::SN_LENGTH;
+
+pub trait SeedBuf {
+    fn as_bytes(&self) -> &[u8; SEED_LENGTH];
+}
+
+#[derive(Zeroize)]
+#[zeroize(drop)]
+pub struct Cdi([u8; SEED_LENGTH]);
+
+impl SeedBuf for Cdi {
+    fn as_bytes(&self) -> &[u8; SEED_LENGTH] {
+        &self.0
+    }
+}
+
+impl Cdi {
+    const REG_ADDR_NONSEC: u32 = 0x40000900;
+
+    // Read the CDI from registers & return to the caller.
+    // Clear the CDI after it's consumed to prevent future access.
+    pub fn new() -> Option<Self> {
+        let mut cdi = [0u8; SEED_LENGTH];
+        let mut offset: usize = 0;
+        let mut addr = Self::REG_ADDR_NONSEC;
+
+        let step = mem::size_of::<u32>();
+        // read out CDI
+        while offset < cdi.len() {
+            let p = addr as *const u32;
+            // core::ptr::read is unsafe by definition.
+            // Support from the lpc55-pac would help:
+            // https://github.com/lpc55/lpc55-pac/issues/15
+            // https://github.com/lpc55/lpc55-pac/pull/16
+            let n = unsafe { ptr::read_volatile(p) };
+
+            cdi[offset..offset + step].copy_from_slice(&n.to_ne_bytes());
+
+            offset += step;
+            addr += step as u32;
+        }
```

Comment:
Untested, but I think something like this would work (with the caveat that I'm not sure why `ptr::read_volatile(p)` was used instead of `ptr::read(p)`, and if that's important this needs tweaking):

```rust
let mut cdi = [0u8; SEED_LENGTH];
// TODO document why this is safe
let source = unsafe {
    core::slice::from_raw_parts(
        Self::REG_ADDR_NONSEC as *const u32,
        SEED_LENGTH / mem::size_of::<u32>(),
    )
};

for (dst, src) in cdi.chunks_mut(mem::size_of::<u32>()).zip(source) {
    dst.copy_from_slice(&src.to_ne_bytes());
}
```

Reply:
Great feedback and thanks for the code snip. I hadn't discovered `core::slice::from_raw_parts` till I had to turn the Hubris image into a slice for hashing. I didn't see it's applicability here but this makes perfect sense. I've made a similar change to the `clear_register` function as well.